language-support/java/bindings-rxjava/src/main/java/com/daml/ledger/rxjava/grpc/CommandsBuilder.java

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.rxjava.grpc;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static java.util.Optional.empty;
+
+import com.daml.ledger.javaapi.data.codegen.HasCommands;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+/**
+ * This class should be used to build a valid submission. It provides {@link #create(String, String, List)}
+ * for initial creation and extra builder methods for additional parameters
+ * e.g {@link #withActAs(List)}, {@link #withWorkflowId(String)} e.t.c
+ *
+ * Usage:
+ * <pre>
+ *   var params = CommandsBuilder.create("applicationId", "commandId", commands)
+ *                                   .withAccessToken("token")
+ *                                   .withParty("party")
+ *                                   .with...
+ * <pre/>
+ */
+public class CommandsBuilder {
+  private String applicationId;
+  private String commandId;
+  private List<@NonNull ? extends HasCommands> commands;
+
+  private String workflowId;
+  private List<@NonNull String> actAs;
+  private List<@NonNull String> readAs;
+  private Optional<Instant> minLedgerTimeAbs;
+  private Optional<Duration> minLedgerTimeRel;
+  private Optional<Duration> deduplicationTime;
+  private Optional<String> accessToken;
+
+  private CommandsBuilder(
+      String workflowId,
+      String applicationId,
+      String commandId,
+      List<@NonNull String> actAs,
+      List<@NonNull String> readAs,
+      Optional<Instant> minLedgerTimeAbs,
+      Optional<Duration> minLedgerTimeRel,
+      Optional<Duration> deduplicationTime,
+      List<@NonNull ? extends HasCommands> commands,
+      Optional<String> accessToken) {
+    this.workflowId = workflowId;
+    this.applicationId = applicationId;
+    this.commandId = commandId;
+    this.actAs = actAs;
+    this.readAs = readAs;
+    this.minLedgerTimeAbs = minLedgerTimeAbs;
+    this.minLedgerTimeRel = minLedgerTimeRel;
+    this.deduplicationTime = deduplicationTime;
+    this.commands = commands;
+    this.accessToken = accessToken;
+  }
+
+  public static CommandsBuilder create(
+      String applicationId, String commandId, List<@NonNull ? extends HasCommands> commands) {
+    return new CommandsBuilder(
+        "",
```

Comment:
empty string or null?

Reply:
Would prefer if users just have to specify a `workflowId`. (they can choose "" if they really wanted to). `null` is not a good idea.

Reply:
Ah it is optional I read. So make it into an `Optional` field, which is not needed as arg to create, but have it as a `withWorkflowId`

Reply:
Yeah I have a `withWorkFlowId`. Making `workFlowId` Optional here though will mean refactoring all the `toProto` functions in `SubmitAndWaitRequest` & `SubmitCommandsRequest` to take the new type `CommandsSubmission`. Not a bad thing, I'll do that